# Лекція 6.1 Базові елементи високорівневої мови програмування.

Програма мовою Python складається з набору інструкцій. Кожна інструкція поміщається на новий рядок. Наприклад:

In [ ]:
print(2 + 3)
print("Привіт")

5
Привіт


Велику роль у Python відіграють відступи. Неправильно поставлений відступ фактично є помилкою. Наприклад, у такому випадку ми отримаємо помилку, хоча код буде практично аналогічний наведеному вище:

In [ ]:
print(2 + 3)
print("Привіт")

5
Привіт


Тому варто поміщати нові інструкції спочатку рядка. У цьому одна з важливих відмінностей Python від інших мов програмування, як-от C# чи Java.

Однак варто враховувати, що деякі конструкції мови можуть складатися з кількох рядків. Наприклад, умовна конструкція if:

In [ ]:
if 1 == 2:
    print("Hello, World!")
print("Hello")

Hello


У цьому випадку якщо 1 менше 2, то виводиться рядок "Hello". І тут уже має бути відступ, оскільки інструкція print("Hello") використовується не сама по собі, а як частина умовної конструкції if. Причому відступ, згідно з інструкцією з оформлення коду, бажано робити з такої кількості пропусків, що кратна 4 (тобто 4, 8, 16 і т. д.) Хоча якщо відступів буде не 4, а 5, то програма також працюватиме.

Таких конструкцій не так багато, тому особливої плутанини з приводу того, де треба, а де не треба ставити пробіли, не повинно виникнути.
Регістрозалежність

Python - регістрозалежна мова, тому вирази print і Print або PRINT представляють різні вирази. І якщо замість методу print для виведення на консоль ми спробуємо використати метод Print:

In [ ]:
print("Привіт")

Привіт


то в нас нічого не вийде.

# Коментарі

Для позначки, що робить та чи інша ділянка коду, застосовуються коментарі. Під час трансляції та виконання програми інтерпретатор ігнорує коментарі, тому вони не мають жодного впливу на роботу програми. Коментарі в Python бувають блокові та рядкові.

Рядкові коментарі випереджаються знаком решітки - #. Вони можуть розташовуватися на окремому рядку:

In [ ]:
# Виведення на консоль
# повідомлення Hello World
print("Hello World")

Hello World


Будь-який набір символів після знака # представляє коментар. Тобто в прикладі вище перші два рядки коду є коментарями.

Також вони можуть розташовуватися на тому самому рядку, що й інструкції мови, після виконуваних інструкцій:

In [ ]:
print("Hello World")  # Виведення повідомлення на консоль

Hello World


У блокових коментарях до і після тексту коментаря ставляться три одинарні лапки: '''текст коментаря'''. Наприклад:

In [ ]:
'''
    Виведення на консоль
    повідомлення Hello World
'''
print("Hello World")

Hello World


# Основні функції

Python надає низку вбудованих функцій. Деякі з них використовуються дуже часто, особливо на початкових етапах вивчення мови, тому розглянемо їх.

Основною функцією для виведення інформації на консоль є функція print(). Як аргумент у цю функцію передається рядок, який ми хочемо вивести:

In [ ]:
print("Hello Python")

Hello Python


Якщо ж нам необхідно вивести кілька значень на консоль, то ми можемо передати їх у функцію print через кому:

In [ ]:
print("Full name:", "Tom", "Smith")

Full name: Tom Smith


У підсумку всі передані значення склеяться через пробіли в один рядок.

Якщо функція print відповідає за виведення, то функція input відповідає за введення інформації. Як необов'язковий параметр ця функція приймає запрошення до введення і повертає введений рядок, який ми можемо зберегти у змінну:

In [ ]:
name = input("Введіть ім`я: ")
print("Привіт,", name)

print(type(name))

Введіть ім`я: Андрій
Привіт, Андрій
<class 'str'>


#Змінні та типи даних
##Змінні

Змінні призначені для зберігання даних. Назва змінної в Python має починатися з алфавітного символу або зі знака підкреслення і може містити алфавітно-цифрові символи та знак підкреслення. І крім того, назва змінної не повинна збігатися з назвою ключових слів мови Python. Ключових слів не так багато, їх легко запам'ятати.

Наприклад, створимо змінну:

In [ ]:
name = "Tom"

Тут визначено змінну name, яка зберігає рядок "Tom".

У Python застосовується два типи найменування змінних: camel case і underscore notation.

Camel case передбачає, що кожне нове підслово в найменуванні змінної починається з великої літери. Наприклад:

In [ ]:
userName = "Tom"

Underscore notation передбачає, що підслова в найменуванні змінної розділяються знаком підкреслення. Наприклад:

In [ ]:
user_name = "Tom"

І також треба враховувати регістрозалежність, тому змінні name і Name представлятимуть різні об'єкти.

In [ ]:
# дві різні змінні
name = "Tom"
Name = "Tom"

Визначивши змінну, ми можемо використовувати в програмі. Наприклад, спробувати вивести її вміст на консоль за допомогою вбудованої функції print:

In [ ]:
name = "Tom"  # визначення змінної name
print(name)   # виведення значення змінної name на консоль

Tom


Відмінною особливістю змінної є те, що ми можемо змінювати її значення протягом роботи програми:

In [ ]:
name = "Tom"  # змінна name дорівнює "Tom"
print(name)   # виводить: Tom
name = "Bob"  # змінюємо значення на "Bob"
print(name)   # виводить: Bob

Tom
123123


## Типи даних

Змінна зберігає дані одного з типів даних. У Python існує безліч різних типів даних. У цьому випадку розглянемо тільки базові типи: bool, int, float, complex і str.
Логічні значення

Тип bool представляє два логічних значення: True (вірно, істина) або False (невірно, брехня). Значення True служить для того, щоб показати, що щось істинне. Тоді як значення False, навпаки, показує, що щось неправдиве. Приклад змінних цього типу:

In [ ]:
isMarried = False
print(isMarried)    # False

isAlive = True
print(isAlive)      # True

False
True


## Цілі числа

Тип int представляє ціле число, наприклад, 1, 4, 8, 50. Приклад

In [ ]:
age = 21
print("Вік:", age)    # Вік: 21

count = 15
print("Кількість:", count) # Кількість: 15

Вік: 21
Кількість: 15


За замовчуванням стандартні числа розцінюються як числа в десятковій системі. Але Python також підтримує числа в двійковій, вісімковій і шістнадцятковій системах.

Для вказівки, що число представляє двійкову систему, перед числом ставиться префікс 0b:

In [ ]:
a = 0b11
b = 0b1011
c = 0b100001
print(a)    # 3 у двійковій системі
print(b)    # 11 у двійковій системі
print(c)    # 33 у двійковій системі

3
11
33


Для вказівки, що число представляє вісімкову систему, перед числом ставиться префікс 0o:

Для вказівки, що число представляє шістнадцяткову систему, перед числом ставиться префікс 0x:

Варто зазначити, що в якій би системі ми не передали число у функцію print для виведення на консоль, воно за замовчуванням буде виводитися в десятковій системі.
##Дробові числа

Тип float представляє число з плаваючою крапкою, наприклад, 1.2 або 34.76. Як роздільник цілої та дробової частин використовується крапка.

In [ ]:
height = 1.68
pi = 3.14
weight = 68.
print(height)   # 1.68
print(pi)       # 3.14
print(weight)   # 68.0

1.68
3.14
68.0


Число з плаваючою крапкою можна визначати в експоненціальному записі:

In [ ]:
x = 3.9e3
print(x)  # 3900.0

x = 3.9e-3
print(x)  # 0.0039

3900.0
0.0039


Число float може мати тільки 18 значущих символів. Так, у цьому випадку використовуються тільки два символи - 3.9. І якщо число занадто велике або занадто мале, то ми можемо записувати число в подібній нотації, використовуючи експоненту. Число після експоненти вказує ступінь числа 10, на яке треба помножити основне число - 3.9.
Комплексні числа

Тип complex представляє комплексні числа у форматі дійсна_частина+уявна_частинаj - після уявної частини вказується суфікс j

In [ ]:
complexNumber = 1+2j
print(complexNumber)   # (1+2j)

(1+2j)


# Рядки

Тип str представляє рядки. Рядок являє собою послідовність символів, укладену в одинарні або подвійні лапки, наприклад "hello" і 'hello'. У Python 3.x рядки представляють набір символів у кодуванні Unicode

In [ ]:
message = "Hello World!"
print(message)  # Hello World!

surname = 'Nikitenko'
print(surname)

name = 'Tom'
print(name)  # Tom

Hello World!
Nikitenko
Tom


При цьому, якщо рядок має багато символів, його можна розбити на частини і ці частини розмістити на різних рядках коду. У цьому випадку весь рядок укладається в круглі дужки, а його окремі частини - в лапки:

In [ ]:
text = ("Laudate omnes gentes laudate "
        "Magnificat in secula ")
print(text)

Laudate omnes gentes laudate Magnificat in secula 


Якщо ж ми хочемо визначити багаторядковий текст, то такий текст укладається в потрійні подвійні або одинарні лапки:

In [ ]:
'''
Це коментар
'''
text = '''Laudate omnes gentes laudate
Magnificat in secula
Et anima mea laudate
Magnificat in secula
'''
print(text)

Laudate omnes gentes laudate
Magnificat in secula
Et anima mea laudate
Magnificat in secula



При використанні потрійних одинарних лапок не варто плутати їх з
коментарями: якщо текст у потрійних одинарних лапках присвоюється
змінній, то це рядок, а не коментар.

## Керуючі послідовності в рядку

Рядок може містити низку спеціальних символів - керуючих послідовностей. Деякі з них:

\\\: дозволяє додати всередину рядка слеш

\': дозволяє додати всередину рядка одинарну лапку

\": дозволяє додати всередину рядка подвійну лапку

\n: здійснює перехід на новий рядок

\t: додає табуляцію (4 відступи)

Застосуємо кілька послідовностей:

In [ ]:
text = "Message:\n\"Hello World\""
print(text)

Message:
"Hello World"


Хоча подібні послідовності можуть нам допомогти в деяких справах, наприклад, помістити в рядок лапку, зробити табуляцію, перенесення на інший рядок. Але вони також можуть і заважати. Наприклад:

In [ ]:
path = "C:\python\name.txt"
print(path)

C:\python
ame.txt


Тут змінна path містить деякий шлях до файлу. Однак усередині рядка зустрічаються символи "\n", які будуть інтерпретовані як керівна послідовність. Так, ми отримаємо такий консольний висновок:

Щоб уникнути подібної ситуації, перед рядком ставиться символ r

In [ ]:
path = r"C:\python\name.txt"
print(path)

C:\python\name.txt


## Вставка значень у рядок

Python дозволяє вставляти в рядок значення інших змінних. Для цього всередині рядка змінні розміщуються у фігурних дужках {}, а перед усім рядком ставиться символ f:

In [ ]:
userName = "Tom"
userAge = 37
user = f"Ім`я: {userName}  вік: {userAge}"
print(user)   # Ім`я: Tom  вік: 37

Ім`я: Tom  вік: 37


У цьому випадку на місце {userName} буде вставлятися значення змінної userName. Аналогічно на замість {userAge} буде вставлятися значення змінної userAge.
Динамічна типізація

Python є мовою з динамічною типізацією. А це означає, що змінна не прив'язана жорстко до певного типу.

Тип змінної визначається виходячи зі значення, яке їй присвоєно. Так, при присвоєнні рядка в подвійних або одинарних лапках змінна має тип str. При присвоєнні цілого числа Python автоматично визначає тип змінної як int. Щоб визначити змінну як об'єкт float, їй присвоюється дробове число, в якому роздільником цілої і дробової частини є крапка.

При цьому в процесі роботи програми ми можемо змінити тип змінної, присвоївши їй значення іншого типу:

In [ ]:
userId = "abc"  # тип str
print(userId)

userId = 234  # тип int
print(userId)

abc
234


За допомогою вбудованої функції type() динамічно можна дізнатися поточний тип змінної:

In [ ]:
userId = "abc"      # тип str
print(type(userId)) # <class 'str'>

userId = 234        # тип int
print(type(userId)) # <class 'int'>

<class 'str'>
<class 'int'>


#Консольне введення та виведення

##Виведення на консоль

Для виведення інформації на консоль призначена вбудована функція print(). Під час виклику цієї функції їй у дужках передається значення, що виводиться:

In [ ]:
print("Hello, Andrii")

Hello, Andrii


Цей код виведе нам на консоль рядок "Hello, Andrii".

Відмінною особливістю цієї функції є те, що за замовчуванням вона виводить значення на окремому рядку. Наприклад:

In [ ]:
print("Hello World")
print("Hello, Andrii")
print("Hello Python")

Hello World
Hello, Andrii
Hello Python


Тут три виклики функції print() виводять деяке повідомлення. Причому під час виведення на консоль кожне повідомлення розміщуватиметься на окремому рядку.

Така поведінка не завжди зручна. Наприклад, ми хочемо, щоб усі значення виводилися на одному рядку. Для цього нам треба налаштувати поведінку функції за допомогою параметра end. Цей параметр задає символи, які додаються наприкінці до виведеного рядка і . При застосуванні параметра end виклик функції print() має такий вигляд: ```print(значення, end = кінцеві_символи)```

За замовчуванням end дорівнює символу "\n", який задає переведення на наступний рядок. Власне тому функція print за замовчуванням виводить значення, що їй передається, на окремому рядку.

Тепер визначимо, щоб функція не робила переведення на наступний рядок, а виводила значення на тому самому рядку:

In [ ]:
print("Hello World", end=" ")
print("Hello, Andrii", end=" ")
print("Hello Python")

Hello World Hello, Andrii Hello Python


Тобто тепер виведені значення розділятимуться пропуском.


In [ ]:
print("Hello World", end=" and ")
print("Hello, Andrii", end=" and ")
print("Hello Python")

Hello World and Hello, Andrii and Hello Python


У цьому разі виведені повідомлення будуть відокремлюватися символами " and ":

## Консольне введення

Поряд із виведенням на консоль ми можемо отримувати введення користувача з консолі, отримувати дані, що вводяться. Для цього в Python визначена функція input(). У цю функцію передається запрошення до введення. А результат введення ми можемо зберегти у змінну. Наприклад, визначимо код для введення користувачем імені:

In [ ]:
name = input("Введіть своє ім'я: ")
print(f"Ваше ім'я: {name}")

Введіть своє ім'я: Андрій
Ваше ім'я: Андрій


У цьому випадку у функцію input() передається запрошення до введення у вигляді рядка "Введіть своє ім'я: ". Результат функції - результат введення користувача передається у змінну name. Потім ми можемо вивести значення цієї змінної на консоль за допомогою функції print(). Приклад роботи коду:

In [ ]:
name = input("Ваше ім'я: ")
age = input("Ваш вік: ")
print(f"Ім'я: {name} Вік: {age}")

Ваше ім'я: Андрій
Ваш вік: 32
Ім'я: Андрій Вік: 32


Варто враховувати, що всі введені значення розглядаються як значення типу str, тобто рядки. І навіть якщо ми вводимо число, як у другому випадку в коді вище, то Python все одно розглядатиме введене значення як рядок, а не як число.

#Арифметичні операції з числами

Python підтримує всі поширені арифметичні операції:


In [ ]:
print(6 + 2)  # 8
print(6 - 2)  # 4
print(6 * 2)  # 12
print(6 / 2)  # 3.0
print(7 / 2)  # 3.5
print(7 // 2)  # 3
print(6 ** 2)  # Зводимо число 6 у ступінь 2. Результат - 36
print(7 % 2)  # Отримання залишку від ділення числа 7 на 2. Результат - 1

8
4
12
3.0
3.5
3
36
1


У разі послідовного використання декількох арифметичних операцій їхнє виконання здійснюється відповідно до їхнього пріоритету. На початку виконуються операції з більшим пріоритетом. Пріоритети операцій у порядку убування наведено в такій таблиці.

** Справо наліво

*/ // % Зліва направо

+- Зліва направо

##Арифметичні операції з присвоєнням

Низка спеціальних операцій дозволяють використовувати присвоїти результат операції першому операнду:

In [ ]:
number = 10
number += 5
print(number)  # 15

number -= 3
print(number)  # 12

number *= 4
print(number)  # 48

15
12
48


##Округлення та функція round

При операціях з числами типу float треба враховувати, що результат операцій з ними може бути не зовсім точним. Наприклад:

In [ ]:
first_number = 2.0001
second_number = 5
third_number = first_number / second_number
print(third_number) # 0.40002000000000004

0.40002000000000004


У цьому випадку ми очікуємо отримати число 0.40002, однак наприкінці через низку нулів з'являється ще якась четвірка. Або ще один вираз:

In [ ]:
print(2.0001 + 0.1)  # 2.1001000000000003

2.1001000000000003


У випадку вище для округлення результату ми можемо використовувати вбудовану функцію round():

In [ ]:
first_number = 2.0001
second_number = 0.1
third_number = first_number + second_number
print(round(third_number))  # 2

2


У функцію round() передається число, яке треба округлити. Якщо у функцію передається одне число, як у прикладі вище, то воно округляється до цілого.

Функція round() також може приймати друге число, яке вказує, скільки знаків після коми повинно містити одержуване число:

In [ ]:
first_number = 2.0001
second_number = 0.1
third_number = first_number + second_number
print(round(third_number, 4))  # 2.1001

2.1001


У цьому випадку число third_number округляється до 4 знаків після коми.

Якщо у функцію передається тільки одне значення - тільки число, що округляється, воно округляється до найближчого цілого

Приклади округлень:

In [ ]:
# округлення до цілого числа
print(round(2.49))  # 2 - округлення до найближчого цілого 2
print(round(2.51))  # 3

2
3


Однак якщо округлювана частина дорівнює однаково віддалена від двох цілих чисел, то округлення йде до найближчого парного:

In [ ]:
print(round(2.5))   # 2 - найближче парне
print(round(3.5))   # 4 - найближче парне

2
4


Округлення проводиться до найближчого кратного 10 у ступені мінус частина, що округляється:

In [ ]:
# округлення до двох знаків після коми
print(round(2.554, 2))      # 2.55
print(round(2.5551, 2))      # 2.56
print(round(2.554999, 2))   # 2.55
print(round(2.499, 2))      # 2.5

2.55
2.56
2.55
2.5


Однак слід враховувати, що функція round() не ідеальний інструмент. Наприклад, вище під час округлення до цілих чисел застосовується правило, згідно з яким, якщо частина, що округляється, однаково віддалена від двох значень, то округлення проводиться до найближчого парного значення. У Python у зв'язку з тим, що десяткова частина числа не може бути точно представлена у вигляді числа float, то це може призводити до деяких не зовсім очікуваних результатів. Наприклад:

In [ ]:
# округлення до двох знаків після коми
print(round(2.545, 2))   # 2.54
print(round(2.555, 2))   # 2.56 - округлення до парного
print(round(2.565, 2))   # 2.56
print(round(2.575, 2))   # 2.58

print(round(2.655, 2))   # 2.65 - округлення не до парного
print(round(2.665, 2))   # 2.67
print(round(2.675, 2))   # 2.67

2.54
2.56
2.56
2.58
2.65
2.67
2.67


#Умовні вирази

Низка операцій представляють умовні вирази. Усі ці операції приймають два операнди і повертають логічне значення, яке в Python представляє тип bool. Існує тільки два логічних значення - True (вираз істинний) і False (вираз хибний).
Операції порівняння

Найпростіші умовні вирази представляють операції порівняння, які порівнюють два значення. Приклади:

In [ ]:
a = 5
b = 6
result = 5 == 6  # зберігаємо результат операції у змінну
print(result)  # False - 5 не дорівнює 6
print(a != b)  # True
print(a > b)  # False - 5 менше 6
print(a < b)  # True

bool1 = True
bool2 = False
print(bool1 == bool2)  # False - bool1 не дорівнює bool2

False
True
False
True
False


Операції порівняння можуть порівнювати різні об'єкти - рядки, числа, логічні значення, однак обидва операнди операції повинні представляти один і той самий тип.
## Логічні операції

Для створення складних умовних виразів застосовуються логічні операції. У Python є такі логічні оператори:

Оператор and (логічне множення) застосовується до двох операндів:
```x and y```
Спочатку оператор and оцінює вираз x, і якщо він дорівнює False, то повертається його значення. Якщо воно дорівнює True, то оцінюється другий операнд - y і повертається значення y.

In [ ]:
age = 20
weight = 58
result = age > 21 and weight == 58
print(result)  # True

False


У цьому випадку оператор and порівнює результати двох виразів: age > 21 weight == 58. І якщо обидва ці вирази повертають True, то оператор and також повертає True (формально повертається значення останнього операнда).

Але операндами оператора and необов'язково виступають значення True і False. Це можуть бути будь-які значення. Наприклад:

In [ ]:
result = 4 and "w"
print(result)  # wоскільки 4 дорівнює True, тому повертається значення останнього операнда

result = 0 and "w"
print(result)  # 0, оскільки 0 еквівалентно False

w
0


У цьому випадку число 0 і порожній рядок "" розглядаються як False, усі інші числа і непорожні рядки еквівалентні True

or (логічне додавання) також застосовується до двох операндів:
```x or y```
Спочатку оператор or оцінює вираз x, і якщо він дорівнює True, то повертається його значення. Якщо він дорівнює False, то оцінюється другий операнд - y і повертається значення y. Наприклад

In [ ]:
age = 22
isMarried = False
result = age > 21 or isMarried
print(result)  # True, оскільки вираз age > 21 дорівнює True

True


І також оператор or може застосовуватися до будь-яких значень. Наприклад:

In [ ]:
result = 4 or "w"
print(result)  #4, оскільки 4 еквівалентно True, тому повертається значення першого операнда

result = 0 or "w"
print(result)  # w, оскільки 0 еквівалентно False, тому повертається значення останнього операнда

4
w


not (логічне заперечення)

Повертає True, якщо вираз дорівнює False

In [ ]:
age = 22
isMarried = False
print(not age > 21)  # False
print(not isMarried)  # True
print(not 4)  # False
print(not 0)  # True

False
True
False
True


##Оператор in

Оператор in повертає True якщо в деякому наборі значень є певне значення. Він має таку форму: ```значення in набір_значень```

In [ ]:
message = "hello world!"
hello = "hello"
print(hello in message)  # True - підрядок hello є в рядку "hello world!"

gold = "gold"
print(gold in message)  # False - підрядка "gold" немає в рядку "hello world!"

True
False


Якщо нам треба навпаки перевірити, чи немає в наборі значень якогось значення, то ми може використовувати модифікацію оператора - not in. Вона повертає True, якщо в наборі значень НЕМАЄ певного значення:

In [ ]:
message = "hello world!"
hello = "hello"
print(hello not in message)  # False

gold = "gold"
print(gold not in message)  # True

False
True


# Умовна конструкція if

Умовні конструкції використовують умовні вирази і залежно від їхнього значення спрямовують виконання програми одним зі шляхів. Одна з таких конструкцій - це конструкція if.У найпростішому вигляді після ключового слова if йде логічний вираз. І якщо цей логічний вираз повертає True, то виконується наступний блок інструкцій, кожна з яких має розпочинатися з нового рядка і має мати відступи від початку виразу if (відступ бажано робити в 4 пропуски або ту кількість пропусків, яка кратна 4):

In [ ]:
language = "english"
if language == "englis":
    print("Hello")
print("End")

End


# Блок else

Якщо раптом нам треба визначити альтернативне рішення на той випадок, якщо вираз в if поверне False, то ми можемо використовувати блок else:

In [ ]:
language = "ukrainian"
if language == "english":
    print("Hello")
else:
    print("Привіт")
print("End")

Привіт
End


Якщо вираз language == "english" повертає True, то виконується блок if, інакше виконується блок else. І оскільки в даному випадку умова language == "english" повертає False, то будуть виконуватися інструкція з блоку else.

Причому інструкції блоку else також повинні мати відступи від початку рядка. Наприклад, у прикладі вище print("End") не має відступу, тому вона не входить до блоку else і буде виконуватися незалежно від того, чому дорівнює умова language == "english". Тобто консоль нам виведе такі рядки:

Блок else також може мати кілька інструкцій, які повинні мати відступ від початку рядка:

In [ ]:
language = "ukrainian"
if language == "english":
    print("Hello")
    print("World")
else:
    print("Привіт")
    print("світ")

Привіт
світ


##elif

Якщо необхідно ввести кілька альтернативних умов, то можна використовувати додаткові блоки elif, після якого йде блок інструкцій.

In [ ]:
language = "german"
if language == "english":
    print("Hello")
    print("World")
elif language == "german":
    print("Hallo")
    print("Welt")
else:
    print("Привіт")
    print("світ")

Hallo
Welt


Спочатку Python перевіряє вираз if. Якщо він дорівнює True, то виконуються інструкції з блоку if. Якщо ця умова повертає False, то Python перевіряє вираз з elif.

Якщо вираз після elif дорівнює True, то виконуються інструкції з блоку elif. Але якщо він дорівнює False то виконуються інструкції з блоку else

За потреби можна визначити кілька блоків elif для різних умов. Наприклад:

In [ ]:
language = "german"
if language == "english":
    print("Hello")
elif language == "german":
    print("Hallo")
elif language == "french":
    print("Salut")
else:
    print("Привіт")

Hallo


## Вкладені конструкції if

Конструкція if зі свого боку сама може мати вкладені конструкції if:

In [ ]:
language = "english"
daytime = "morning"
if language == "english":
    print("English")
    if daytime == "morning":
        print("Good morning")
    else:
        print("Good evening")

English
Good morning


#Цикли
Цикли дають змогу виконувати деяку дію залежно від дотримання певної умови. У мові Python є такі типи циклів: while, for

##Цикл while

Цикл while перевіряє істинність деякої умови, і якщо умова істинна, то виконує інструкції циклу.Після ключового слова while вказується умовний вираз, і доки цей вираз повертає значення True, виконуватиметься блок інструкцій, що йде далі.

Усі інструкції, які відносяться до циклу while, розташовуються на наступних рядках і повинні мати відступ від початку ключового слова while.

In [ ]:
number = 1
flag = True

while flag:
    print(f"number = {number}")
    number += 1
    if number == 3:
        flag = False
print("Роботу програми завершено")

number = 1
number = 2
Роботу програми завершено


У цьому випадку цикл while виконуватиметься, поки змінна number менша за 5.

Сам блок циклу складається з двох інструкцій:

In [ ]:
print(f"number = {number}")
number += 1

number = 3


Зверніть увагу, що вони мають відступи від початку оператора while - у цьому випадку від початку рядка. Завдяки цьому Python може визначити, що вони належать циклу. У самому циклі спочатку виводиться значення змінної number, а потім їй присвоюється нове значення.

Також зверніть увагу, що остання інструкція print("Робота програми завершена") не має відступів від початку рядка, тому вона не входить до циклу while.

Увесь процес циклу можна представити таким чином:

Спочатку перевіряється значення змінної number - чи менше воно 5. І оскільки спочатку змінна дорівнює 1, то ця умова повертає True, і тому виконуються інструкції циклу

Інструкції циклу виводять на консоль рядок number = 1. І далі значення змінної number збільшується на одиницю - тепер вона дорівнює 2. Одноразове виконання блоку інструкцій циклу називається ітерацією. Тобто таким чином, у циклі виконується перша ітерація.

Знову перевіряється умова number < 5. Вона, як і раніше, дорівнює True, оскільки number = 2, тому виконуються інструкції циклу

Інструкції циклу виводять на консоль рядок number = 2. І далі значення змінної number знову збільшується на одиницю - тепер вона дорівнює 3. Таким чином, виконується друга ітерація.

Знову перевіряється умова number < 5. Вона, як і раніше, дорівнює True, оскільки number = 3, тому виконуються інструкції циклу

Інструкції циклу виводять на консоль рядок number = 3. І далі значення змінної number знову збільшується на одиницю - тепер вона дорівнює 4. Тобто виконується третя ітерація.

Знову перевіряється умова number < 5. Вона, як і раніше, дорівнює True, оскільки number = 4, тому виконуються інструкції циклу

Інструкції циклу виводять на консоль рядок number = 4. І далі значення змінної number знову збільшується на одиницю - тепер вона дорівнює 5. Тобто виконується четверта ітерація.

І знову перевіряється умова number < 5. Але тепер вона дорівнює False, оскільки number = 5, тому виконується вихід із циклу. Усе цикл - завершився. Далі вже виконуються дії, які визначені після циклу. Таким чином, цей цикл зробить чотири проходи або чотири ітерації

Для циклу while також можна визначити додатковий блок else, інструкції якого виконуються, коли умова дорівнює False:

In [ ]:
number = 1

while number < 5:
    print(f"number = {number}")
    number += 1
else:
    print(f"number = {number}. Роботу циклу завершено")
print("Роботу програми завершено")

number = 1
number = 2
number = 3
number = 4
number = 5. Роботу циклу завершено
Роботу програми завершено


Тобто в цьому випадку спочатку перевіряється умова і виконуються інструкції while. Потім, коли умова стає рівною False, виконуються інструкції з блоку else. Зверніть увагу, що інструкції з блоку else також мають відступи від початку конструкції циклу.

##Цикл for

Інший тип циклів представляє конструкція for. Цей цикл пробігається по набору значень, поміщає кожне значення у змінну, і потім у циклі ми можемо з цією змінною виконувати різні дії. Після ключового слова for йде назва змінної, в яку будуть поміщатися значення. Потім після оператора in вказується набір значень і двокрапка.

А з наступного рядка розташовується блок інструкцій циклу, які також повинні мати відступи від початку циклу.

Під час виконання циклу Python послідовно отримує всі значення з набору і передає їх змінній. Коли всі значення з набору будуть перебрані, цикл завершує свою роботу.

Як набір значень, наприклад, можна розглядати рядок, який по суті являє собою набір символів. Подивимося на прикладі:

In [ ]:
message = "Hello"

for c in message:
    print(c, end=" ")

H e l l o 

У циклі визначається змінна c, після оператора in в якості набору, що перебирається, вказана змінна message, яка зберігає рядок "Hello". У підсумку цикл for буде перебирати послідовно всі символи з рядка message і поміщати їх у змінну c. Блок самого циклу складається з однієї інструкції, яка виводить значення змінної с на консоль.

##Вкладені цикли

Одні цикли всередині себе можуть містити інші цикли. Розглянемо на прикладі виведення таблиці множення:

In [ ]:
i = 1
j = 1
while i < 10:
    while j < 10:
        print(i * j, end="\t")
        j += 1
    print("\n")
    j = 1
    i += 1

1	2	3	4	5	6	7	8	9	

2	4	6	8	10	12	14	16	18	

3	6	9	12	15	18	21	24	27	

4	8	12	16	20	24	28	32	36	

5	10	15	20	25	30	35	40	45	

6	12	18	24	30	36	42	48	54	

7	14	21	28	35	42	49	56	63	

8	16	24	32	40	48	56	64	72	

9	18	27	36	45	54	63	72	81	



Зовнішній цикл while i < 10: спрацьовує 9 разів, доки змінна i не дорівнюватиме 10. Усередині цього циклу спрацьовує внутрішній цикл while j < 10:. Внутрішній цикл також спрацьовує 9 разів, поки змінна j не дорівнюватиме 10. Причому всі 9 ітерацій внутрішнього циклу спрацьовують у рамках однієї ітерації зовнішнього циклу.

У кожній ітерації внутрішнього циклу на консоль буде виводитися добуток чисел i і j. Потім значення змінної j збільшується на одиницю. Коли внутрішній цикл закінчив роботу, значення змінної j скидається в 1, а значення змінної i збільшується на одиницю і відбувається перехід до наступної ітерації зовнішнього циклу. І все повторюється, поки змінна i не дорівнюватиме 10. Відповідно внутрішній цикл спрацює всього 81 раз для всіх ітерацій зовнішнього циклу.

In [ ]:
for c1 in  "ab":
    for c2 in "ba":
        print(f"{c1}{c2}")

ab
aa
bb
ba


# Вихід із циклу. break і continue

Для керування циклом ми можемо використовувати спеціальні оператори break і continue. Оператор break здійснює вихід із циклу. А оператор continue виконує перехід до наступної ітерації циклу.

Оператор break може використовуватися, якщо в циклі утворюються умови, які несумісні з його подальшим виконанням. Розглянемо такий приклад:

In [ ]:
number = 0
while number < 5:
    number += 1
    if number == 3 :    # якщо number = 3, виходиму з циклу
        break
    print(f"number = {number}")


number = 1
number = 2


Тут цикл while перевіряє умову number < 5. І поки number не дорівнює 5, передбачається, що значення number буде виводитися на консоль. Однак усередині циклу також перевіряється інша умова: if number == 3. Тобто, якщо значення number дорівнює 3, то за допомогою оператора break виходимо з циклу.

На відміну від оператора break оператор continue виконує перехід до наступної ітерації циклу без його завершення. Наприклад, у попередньому прикладі замінимо break на continue:

In [ ]:
number = 0
while number < 5:
    number += 1
    if number == 3 :    # якщо number = 3, переходимо до нової ітерації циклу
        continue
    print(f"number = {number}")

number = 1
number = 2
number = 4
number = 5
